### PR TITLE
[skipruntime] Manage resource sharing

### LINKED ISSUE #522: Re-enable tests for external services
_Originally posted by @jberdine in https://github.com/SkipLabs/skip/pull/515#discussion_r1853970844_

### PR DESCRIPTION
**Pull request by skiplabsdaniel**
Close #522

### CODE REVIEW

File: skipruntime-ts/native/src/Extern.sk [hunk 1]
```diff
@@ -607,7 +607,9 @@ fun getAllOfRuntime(
     }
   | _ ->
     SKStore.runWithResult(context ~> {
-      getAll(context, resource, params(jsonParams), optRequest)
+      res = getAll(context, resource, params(jsonParams), optRequest);
+      context.update();
```
Comment (beauby) [inline skipruntime-ts/native/src/Extern.sk]: Why is this needed now and not before?

File: skipruntime-ts/native/src/Extern.sk [hunk 2]
```diff
@@ -648,7 +650,9 @@ fun getForKeyOfRuntime(
     }
   | _ ->
     SKStore.runWithResult(context ~> {
-      getForKey(context, resource, params(jsonParams), key, optRequest)
+      res = getForKey(context, resource, params(jsonParams), key, optRequest);
+      context.update();
```
Comment (beauby) [inline skipruntime-ts/native/src/Extern.sk]: Same question as above.

File: skipruntime-ts/wasm/package.json [hunk 3]
```diff
@@ -13,7 +13,7 @@
   },
   "devDependencies": {
     "@types/mocha": "^10.0.8",
-    "@skipruntime/tests": "^0.0.3",
+    "@skipruntime/tests": "^0.0.4",
```
Comment (beauby) [inline skipruntime-ts/wasm/package.json]: This should go in a separate commit.

File: skipruntime-ts/wasm/src/internals/skipruntime_module.ts [hunk 4]
```diff
@@ -1243,7 +1243,7 @@ export class ServiceInstance {
   instantiateResource(
     identifier: string,
     resource: string,
-    params: { [param: string]: string },
+    params: { [param: string]: string } = {},
```
Comment (beauby) [inline skipruntime-ts/wasm/src/internals/skipruntime_module.ts]: Is this related to resource sharing? If not let's put it in a different commit.

File: skipruntime-ts/native/src/Runtime.sk [hunk 5]
```diff
@@ -239,16 +247,35 @@ fun buildResourcesGraph(
...
     ResourceInfo::type,
     context,
     kResourceCollectionsDir,
-    (context, writer, key, it) ~> {
+    (context, writer, key, _it) ~> {
       pushContext(context);
       try {
-        resourceId = base64(key.value);
-        def = it.first;
+        resourceId = toResoureId(key.name, key.params.value);
```
Comment (beauby) [inline skipruntime-ts/native/src/Runtime.sk]:
```suggestion
        resourceId = toResourceId(key.name, key.params.value);
```

File: skipruntime-ts/native/src/Runtime.sk [hunk 6]
```diff
@@ -239,16 +247,35 @@ fun buildResourcesGraph(
...
+    SKStore.IntFile::type,
+    context,
+    dirname.sub("availables"),
+    (_c, writer, key, _it) ~> {
+      // map to ignore array changes
+      // to prevent recompute on each add/remove session
+      writer.set(key, SKStore.IntFile(1))
+    },
+  );
+  _ = availablesHdl.map(
```
Comment (beauby) [inline skipruntime-ts/native/src/Runtime.sk:270]: Why not directly map on `resourcesHdl` and use `it.first`?

File: skipruntime-ts/native/src/Extern.sk (hunk 1 above)
Comment (skiplabsdaniel) [reply skipruntime-ts/native/src/Extern.sk]: I realized that this was missing since the status management, it works without but it would be cleaner with. But it's true that it's not part of this commit.

File: skipruntime-ts/native/src/Extern.sk (hunk 2 above)
Comment (skiplabsdaniel) [reply skipruntime-ts/native/src/Extern.sk]: Same

File: skipruntime-ts/native/src/Runtime.sk (hunk 6 above)
Comment (skiplabsdaniel) [reply skipruntime-ts/native/src/Runtime.sk:270]: Because each source will compute its own graph, it's allow to have have one source force a resource reactive graph.
